rules/DowngradePhp74/Rector/LNumber/DowngradeNumericLiteralSeparatorRector.php

```diff
@@ -83,24 +88,33 @@ public function refactor(Node $node): ?Node
          * by adding ".0" at the end (eg: 0.0).
          * Then, add it again.
          */
-        if ($node instanceof DNumber && ! \str_contains($node->value, '.')) {
-            $node->value .= '.0';
+        if (! \str_contains($numberValueAsString, '.')) {
+            $numberValueAsString .= '.0';
         }
 
-        if (! str_contains($node->value, '_')) {
-            return null;
-        }
+        $node->value = (float) $numberValueAsString;
 
         return $node;
     }
 
-    public function shouldRefactor(LNumber | DNumber $node): bool
+    private function shouldSkip(LNumber | DNumber $node): bool
     {
         // "_" notation can be applied to decimal numbers only
         if ($node instanceof LNumber) {
-            return $node->getAttribute(AttributeKey::KIND) === LNumber::KIND_DEC;
+            $numberKind = $node->getAttribute(AttributeKey::KIND);
+            if ($numberKind !== LNumber::KIND_DEC) {
+                return true;
+            }
+        }
+
+        // we have to hack around tokens to get original value, see https://github.com/nikic/PHP-Parser/pull/832
+        $oldTokens = $this->file->getOldTokens();
+        $tokenValue = $oldTokens[$node->getStartTokenPos()][1] ?? null;
+
+        if ($tokenValue === null) {
+            return true;
         }
 
-        return true;
+        return ! str_contains($tokenValue, '_');
```

Comment:
Hopefully we can get rid of this global token access after raw value is supported nativelly in both `LNumber` and `DNumber`:

* https://github.com/nikic/PHP-Parser/pull/832
* https://github.com/nikic/PHP-Parser/pull/833